Split functional testings via github action matrix

mcmilk opened the pull request:
### Motivation and Context

The functional testimgs are timing out more and more ... the max. time for execution is 360m ... this seems to be reached :(

### Description

This commit changes the workflow of the github actions.

We split the workflow into different parts:

1) build zfs modules for Ubuntu 20.04 and 22.04 (~25m)
2) 2x zloop test (~10m) + 2x sanity test (~25m)
3) functional testings in parts 1..5 (each ~1h)
   - these could be triggered, when sanity tests are ok
   - currently I just start them all in the same time
4) cleanup and create summary

When everything is fine, the full run with all testings
should be done in around 2 hours.

The codeql.yml and checkstyle.yml are not part in this circle.

The testings are also modfied a bit:
- report info about CPU and checksum benchmarks
- reset the debugging logs for each test
  - when some error occurred, we call dmesg with -c to get
    only the log output for the last failed test
  - we empty also the dbgsys

## The testings are done this way

```mermaid
flowchart TB
subgraph CleanUp and Summary
  Part1-20.04-->CleanUp+nice+Summary
  Part2-20.04-->CleanUp+nice+Summary
  PartN-20.04-->CleanUp+nice+Summary
  Part1-22.04-->CleanUp+nice+Summary
  Part2-22.04-->CleanUp+nice+Summary
  PartN-22.04-->CleanUp+nice+Summary
end

subgraph Testings Ubuntu 20.04
  sanity-checks-20.04
  functional-testing-20.04-->Part1-20.04
  functional-testing-20.04-->Part2-20.04
  functional-testing-20.04-->PartN-20.04
  zloop-checks-20.04
end

subgraph Testings Ubuntu 22.04
  sanity-checks-22.04
  functional-testing-22.04-->Part1-22.04
  functional-testing-22.04-->Part2-22.04
  functional-testing-22.04-->PartN-22.04
  zloop-checks-22.04
end

subgraph Code Checking + Building
  codeql.yml
  Build-Ubuntu-20.04-->sanity-checks-20.04
  Build-Ubuntu-22.04-->sanity-checks-22.04
  Build-Ubuntu-20.04-->zloop-checks-20.04
  Build-Ubuntu-22.04-->zloop-checks-22.04
  checkstyle.yml
end
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

gmelikov commented on .github/workflows/zfs-build.yaml:
```diff
@@ -0,0 +1,218 @@
+        os: [18.04, 20.04 ]
+        tests: [ part1, part2, part3, part4, part5 ]
+    runs-on: ubuntu-${{ matrix.os }}
+    needs: tests-sanity
```
`Sanity tests` is a subset from `Functional tests` just to get basic results faster. So I propose to leave them run in parallel
```suggestion
    needs: build
```

mcmilk replied:
The two sanity tests are run in parallel after building the modules and need about 25m.
The idea is, when the sanity tests are failing, we have no need to start the functional tests et all.

gmelikov replied:
> we have no need to start the functional tests et all.

But we can get more failed tests without need to wait for sanity tests' green run next time. Otherwise there will be a one more tests fixing and CI waiting cycle guaranteed for developer. More time to spend - less motivation.

gmelikov replied:
As far as i know we don't have to save any runners' resources of github actions, it's free and we may not want to trade free cpu time for a contributor expensive time, this is my concern.

mcmilk replied:
Good point ... I should adjust the needs flags.

gmelikov replied:
Aaand if we are on a road of splitting tests per parallel jobs, we can exclude `sanity tests` group from full runners after all, but it'll be harder than just change running tags of tests (thought for future)

gmelikov commented on .github/workflows/test-dependencies.txt:
```diff
@@ -0,0 +1,39 @@
+acl
```
Optional for this PR, but - maybe cleanup https://github.com/openzfs/zfs/blob/master/.github/workflows/build-dependencies.txt from test deps and group them when needed?

mcmilk replied:
Yes, the build- and test- dependencies need to be re-checked.

mcmilk replied:
The splitting within `.github/workflows/scripts/split_functional_tests.sh` may also need some update... it's just a start. We will see how the runners differ in time... and can fine tune the things a bit.

gmelikov commented:
[1h 49m](https://github.com/openzfs/zfs/actions/runs/3310524566) looks like a great result for full test suite run even as a start!

mcmilk commented:
> [1h 49m](https://github.com/openzfs/zfs/actions/runs/3310524566) looks like a great result for full test suite run even as a start!

Yes ;-)

I have removed the Ubuntu 22.04 sanity and functional tests, so this PR could maybe merged a bit faster.
The commit message should also be fine now and also some ZFS module cleanup was added as last step.

gmelikov commented on .github/workflows/build-dependencies.txt:
```diff
@@ -1,18 +1,11 @@
 build-essential
-curl
-dbench
+dkms
```
We don't use dkms now, looks like we don't need to add it.

mcmilk replied:
I will try this.

gmelikov commented on .github/workflows/build-dependencies.txt:
```diff
@@ -27,23 +20,11 @@ libselinux-dev
-nfs-kernel-server
-pamtester
-parted
+linux-headers-generic
```
Without dkms we don't need headers too, AFAIK

mcmilk replied:
I will try this.

gmelikov commented:
Overall thoughts:
- speed gain is really a good reason to use it, and we won't get into timeouts in the foreseeable future
- one workflow file makes it harder to read manifest, but we can impove it in future by moving common parts in scripts
- many additional `checks` in PR's status worsens its readability, but looks like there is no other way to parallelize now https://github.com/orgs/community/discussions/26246

andrewc12 commented:
I like this idea of splitting it up like this in a matrix. tried it myself just didn't end up writing a script. :)

mcmilk commented:
> I like this idea of splitting it up like this in a matrix. tried it myself just didn't end up writing a script. :)

We should use this for MacOS and Windows as well... when you have problems with this yaml stuff, I could help a bit.

andrewc12 commented:
> We should use this for MacOS and Windows as well... when you have problems with this yaml stuff, I could help a bit.

Heh, I'm usually fine with that stuff, but honestly if you think you could help with anything I'd be glad to take it.

behlendorf commented:
> we can exclude sanity tests group from full runners after all

I agree, since the sanity tests are just a subset of the full test run there's not much value in running them, unless they're used to gate a full test run.

If we parallelize the tests like this we could also split them up in some meaningful way.  Perhaps "cli tests" (cli_root, cli_user), "feature tests" (initialize, trim, compression, fallocate, quota, send/recv, snapshots, checkpoints, etc), and "resilience tests" (raidz, redundancy, removal, replacement, slog, online_offline, scrub_mirror, etc).  Or some other relatively equal split.  We'd just need to settle on the group names and add them to the `tags` sections of the run file.

My inclination would be to try and split it only 3 or 4 ways to cut down a bit on the number of status reports and try and keep it as readable as possible.

mcmilk commented:
> > we can exclude sanity tests group from full runners after all
> 
> I agree, since the sanity tests are just a subset of the full test run there's not much value in running them, unless they're used to gate a full test run.

Then I will remove it.

> If we parallelize the tests like this we could also split them up in some meaningful way. Perhaps "cli tests" (cli_root, cli_user), "feature tests" (initialize, trim, compression, fallocate, quota, send/recv, snapshots, checkpoints, etc), and "resilience tests" (raidz, redundancy, removal, replacement, slog, online_offline, scrub_mirror, etc). Or some other relatively equal split. We'd just need to settle on the group names and add them to the `tags` sections of the run file.

Okay, i will try that ;)

> 
> My inclination would be to try and split it only 3 or 4 ways to cut down a bit on the number of status reports and try and keep it as readable as possible.

We could also collect each summary of the tags and create a big one in the last step.

mcmilk commented:
> If we parallelize the tests like this we could also split them up in some meaningful way. Perhaps "cli tests" (cli_root, cli_user), "feature tests" (initialize, trim, compression, fallocate, quota, send/recv, snapshots, checkpoints, etc), and "resilience tests" (raidz, redundancy, removal, replacement, slog, online_offline, scrub_mirror, etc). Or some other relatively equal split. We'd just need to settle on the group names and add them to the `tags` sections of the run file.

I created a helper script for modifying the common.run file: https://gist.github.com/mcmilk/acc54a671ad2164ce98a0996a2ad6562

@behlendorf - maybe you want to have some changes on this initial grouping ?

gmelikov commented:
If we go into hardcoding tags, then we may want to at least find non-described tags and run it in one of parts

ryao commented:
Something that we might want to address either as part of this or separately is that when the buildbot runs the ZTS, it gives multiple files. One is called "tests", which is just a summary of which tests passed and which failed. Another is called "log", which shows the details of each test. For example:

```
21:47:09.61 NOTE: Check filesystem other2 readonly testpool/testfs/SUBFS
21:47:09.62 NOTE: other2 zfs set checksum=on testpool/testfs/SUBFS
21:47:09.62 NOTE: user: other2
21:47:09.62 NOTE: cmd: zfs set checksum=on testpool/testfs/SUBFS
```

When a test fails, this is helpful information that can often be used in debugging. With the github workflows, we do not have this information and the results are useless for trying to address the issue. The buildbot is currently failing to run the ZTS, which makes this absence of information from the github workflows particularly noticeable.

mcmilk commented:
@gmelikov @behlendorf - I reworked the Github actions again.

I named this `zfs-linux.yaml` - so we can later add `zfs-windows.yaml` and `zfs-macos.yaml` files.
It's again just sth. like an requst for comment ... but of cause - we need some speed within the testings ;-)

I could also splt this pull request into seperate parts like this:
1. the yaml files and new shell scripts
2. the changes with the testings ... `dmesg -c` and so on ...

andrewc12 commented:
Neat, this is a lot of good ideas.
I have a few comments.
- I like the building only once per os
- I dislike how the sanity/zloop/functional checks are now hidden under a single test, but this might be unavoidable.
- zfs-tests.yml is linux specific, could you rename it to zfs-tests-linux.yml

Speaking of future macos and windows tests, they're not going to be coming to github soon due to various issues.

mcmilk commented:
This small difference makes the module building (job 1) and later installing (testings) a bit bad:

```
Current runner version: '2.302.1'
Runner Image
  Image: ubuntu-20.04
  Version: 20230305.1  vs  Version: 20230224.2
```

Maybe I should switch to dkms .... I am thinking about this...

mcmilk commented on .github/workflows/scripts/generate-summary.sh:
```diff
@@ -0,0 +1,73 @@
+  TEST="$2"
+  echo "::group::Generating summary $LOGDIR"
+  # generate all in one files from the part ones
+  for p in $LOGDIR/part[123456789].tar; do
```
I will use the suggested brace method. So we can check for missing logfiles an give some fatal error report when this happens.

mcmilk commented on .github/workflows/scripts/generate-summary.sh:
```diff
@@ -0,0 +1,73 @@
+ZTS_REPORT="tests/test-runner/bin/zts-report.py"
+chmod +x $ZTS_REPORT
+
+# all output goes to summary page
```
Yes I have, but with the missing logfiles the output will be different ... I will create a new output today and put a link here.

mcmilk replied:
https://github.com/mcmilk/testrepo/blob/master/SUM.md

mcmilk commented on .github/workflows/build-dependencies.txt:
```diff
@@ -2,56 +2,49 @@ acl
 libssl-dev
 libtool
 libudev-dev
-linux-headers-generic
```
No, but compiling was fine. Other kernel header packages were there ...but I put in the  `linux-headers-generic` again.

mcmilk commented:
Okay, the main thing is done now.

Sometimes the functional testings (part3) will fail. This seems unrelated to this functional tests splitting.
Also the old workflows seem to hang sometimes on the slog tests. The test `slog_014_pos (run as root) [00:21]` seems fine, and the next one is timing out ....

I splitted the "Cleanup and summary" into 5 calls -because we have a limit of 1 MiB per step which can be put into the summary. The first call does the summary things and the next 4 calls will print debugging details (`dmesg`, `dbgmsg`) of the failed tests.

mcmilk commented:
This is request for review now?

We went down from `>4h` to `1h 45m` ;-)
We also have a sumary in the end and some first debugging can be done with the debug files in the end. These files are truncated to ~1MB ... but this should be okay for most cases.

behlendorf reviewed (approved):
Having a [summary page](https://github.com/openzfs/zfs/actions/runs/4414917525) where we can easily see the logs from tests which failed is real quality of life improvement!  As are the faster test run times.  I'm good with this, let's go ahead and merge this and refine as needed.  Thanks.

gmelikov commented on .github/workflows/zfs-linux-tests.yml:
```diff
@@ -0,0 +1,113 @@
+          /var/tmp/zloop/*/vdev/
+        if-no-files-found: ignore
+
+  sanity:
```
IIRC we discussed that now we may not even need sanity tests, but this PR is already to good to be true (thanks @mcmilk ) so let it be

gmelikov commented on .github/workflows/zfs-linux.yml:
```diff
@@ -0,0 +1,71 @@
+      with:
+        name: Summary Files
+        path: Summary/
+    - uses: geekyeggo/delete-artifact@v2
```
We may leave artifacts, it may be convenient for someone to download already built debug build-release, but just loud thoughts, never mind for now.

mcmilk replied:
Yes, the generated modules will be deleted, but all logs are saved now.
Currently the logs are only saved, when sth. failed .... this is changed to `we log always`.

gmelikov reviewed (approved):
LGTM, let's try and use it! Great work @mcmilk , thank you!

behlendorf commented:
Merged.  Let's see how it goes and refine as needed.  Thanks!

mcmilk commented:
There is an error, when the images rotate: https://github.com/actions/runner-images#available-images

So currently some runners have 20230313, and some have 20230305 .... I checked this via the file `tests/ImageOS.txt` ... and the module should compile if this applies. But of cause, we didn't checkout the current release with git ... I need to add this.

I will create a pull request for it... sorry, but this issue could not be testet, cause all runners had the same version ...

gmelikov commented:
@mcmilk that was fast to get into this corner case! Glad we can just rerun whole pipeline
